lib/Target/ExportSystemC/EmissionPattern.h

```diff
@@ -0,0 +1,362 @@
+  ADD = 6,
+  SUB = 6,
+  SHL = 7,
+  SHR = 7,
+  RELATIONAL = 9,
+  EQUALITY = 10,
+  BITWISE_AND = 11,
+  BITWISE_XOR = 12,
+  BITWISE_OR = 13,
+  LOGICAL_AND = 14,
+  LOGICAL_OR = 15,
+  TERNARY = 16,
+  THROW = 16,
+  ASSIGN = 16,
+  COMMA = 17
+};
+
+/// This class allows a pattern's match function for inlining to pass its
+/// result's precedence to the pattern that requested the expression.
+class MatchResult {
+public:
+  MatchResult() = default;
+  MatchResult(Precedence precedence)
+      : isFailure(false), precedence(precedence) {}
+
+  bool failed() const { return isFailure; }
+  Precedence getPrecedence() const { return precedence; }
+
+private:
+  bool isFailure = true;
+  Precedence precedence;
+};
+
+/// This class is returned to a pattern that requested inlined emission of a
+/// value. It allows the pattern to emit additional characters before the
+/// requested expression depending on the precedence.
+class InlineEmitter {
+public:
+  InlineEmitter(std::function<void()> emitter, Precedence precedence)
+      : precedence(precedence), emitter(std::move(emitter)) {}
+
+  Precedence getPrecedence() const { return precedence; }
+  void emit() const { emitter(); }
+
+private:
+  Precedence precedence;
+  std::function<void()> emitter;
+};
+
+//===----------------------------------------------------------------------===//
+// Emission pattern base classes.
+//===----------------------------------------------------------------------===//
+
+/// This is indented to be the base class for all emission patterns.
+class PatternBase {
+public:
+  explicit PatternBase(const void *rootValue) : rootValue(rootValue) {}
+
+  template <typename E, typename... Args>
+  static std::unique_ptr<E> create(Args &&...args) {
+    std::unique_ptr<E> pattern =
+        std::make_unique<E>(std::forward<Args>(args)...);
+    return pattern;
+  }
+
+  /// Get a unique identifier for the C++ type the pattern is matching on. This
+  /// could be a specific MLIR type or operation.
+  const void *getRootValue() const { return rootValue; }
+
+private:
+  const void *rootValue;
+};
+
+/// This is intended to be the base class for all emission patterns matching on
+/// operations.
+struct OpEmissionPatternBase : PatternBase {
+  OpEmissionPatternBase(StringRef operationName, MLIRContext *context)
+      : PatternBase(
+            OperationName(operationName, context).getAsOpaquePointer()) {}
+  virtual ~OpEmissionPatternBase() = default;
+
+  /// Checks if this pattern is applicable to the given value and configuration
+  /// to emit an inlinable expression. Additionally returns information such as
+  /// the precedence to the pattern where this pattern's result is to be
+  /// inlined.
+  virtual MatchResult matchInlinable(Value value, EmissionConfig &config) = 0;
+
+  /// Checks if this pattern is applicable to the given operation and
+  /// configuration for statement emission.
+  virtual bool matchStatement(mlir::Operation *op, EmissionConfig &config) = 0;
```

Comment:
This would rely on the pattern implementer to do all the matching logic before the first emission and not cancel afterwards. To support this properly we would need a staging `ostream` that can be discarded when the pattern returns `false`. This might hurt performance more than strictly splitting the matching and emission phase because most patterns have a very simple matching logic and we would write twice to an ostream for nothing most of the time. Not sure if we could efficiently implement an `ostream` than can roll back already written characters to a checkpoint because that would be the less frequent operation.